src/Universum/String/Conversion.hs

```diff
@@ -159,6 +162,96 @@ instance ToString T.Text where
 instance ToString LT.Text where
     toString = LT.unpack
 
+{- [Note toString-toText-rewritting]
+
+@toString . toText@ pattern may occur quite often after inlining because
+we tend to use 'Text' rather than 'String' in function signatures, but
+there are still some libraries which use 'String's and thus make us perform
+conversions back and forth.
+
+Herewith, it's tempting to define the following rule:
+
+@
+{-# RULES T.unpack (T.pack s) = s #-}.
+@
+
+This would make as to mark 'T.unpack' and 'T.pack' as non-inlinable at the
+first phases of simplifier pass because if they get inlined then obviously
+our rule cannot be applied.
+We can go this way, but we can do even better if take rules defined in
+'Data.Text' into account.
+
+Quoting investigation of @int-index:
+
+If we look at @unpack@ and @pack@ they are defined as
+
+@
+unpack = S.unstreamList . stream
+{-# INLINE [1] unpack #-}
+
+pack = unstream . S.map safe . S.streamList
+{-# INLINE [1] pack #-}
+@
+
+After they get inlined, the rule seems to be
+
+@
+(S.unstreamList . stream) ((unstream . S.map safe . S.streamList) a)
+@
+
+If we also inline function composition, we get
+
+@
+S.unstreamList (stream (unstream (S.map safe (S.streamList a))))
+@
+
+`stream` and `unstream` surely cancel out via this rule:
+
+@
+{-# RULES "STREAM stream/unstream fusion" forall s. stream (unstream s) = s #-}
+@
+
+So we are left with
+
+@
+S.unstreamList (S.map safe (S.streamList a))
+@
+
+Now, what's this 'safe' function? Turns out it's defined as
+
+@
+safe :: Char -> Char
+safe c
+    | ord c .&. 0x1ff800 /= 0xd800 = c
+    | otherwise                    = '\xfffd'
+@
+
+Aha, so it's mapping some codepoints to @'\xfffd'@!
+There's a comment on top of it to explain this:
+
+```
+-- UTF-16 surrogate code points are not included in the set of Unicode
+-- scalar values, but are unfortunately admitted as valid 'Char'
+-- values by Haskell.  They cannot be represented in a 'Text'.  This
+-- function remaps those code points to the Unicode replacement
+-- character (U+FFFD, \'&#xfffd;\'), and leaves other code points
+-- unchanged.
+```
+
+This logic is lost with the mentioned rewrite rule.
+Not a huge loss, but it does mean that this rewrite rule isn't meaning preserving.
```

Comment:
I'd like to fully understand consequences of this loss and be more explicit about it. AFAIU, it means that if we have `String` which contains UTF-16 surrogate code point and apply `toString . toText` to it, we'll get different results with and without this rule. Since we ignore `safe`, does it mean that `error` will be called somewhere?
I propose to:
1. Add a test where we generate a `String` with such character and demonstrate that behavior is different. I don't know what exactly that test should do and which behavior is expected, but without such test it's hard to understand in which case and how this rule changes the semantics.
2. Mention it in `Gotchas` in README.

Reply:
1. It's not clear how can I illustrate this behavior. As [comment](http://hackage.haskell.org/package/text-1.2.3.1/docs/src/Data.Text.Internal.html#showText) to `safe` function suggests, `Text` cannot contain surrogate code points, so this `safe` will not change anything if my `String` is constructed from `Text`. Unless my text is constructed incorrectly, the rule does not change the semantics.

2. Added.

Reply:
1. But what if you apply `toString . toText` to `[c]` where `ord c .&. 0x1ff800 == 0xd800`? AFAIU, with this rule and without this rule the result will be different, am I wrong?

Reply:
Right, ya tuplu.